In [1]:
import requests
import pandas as pd
import numpy as np
from datetime import datetime
from tqdm.notebook import tqdm 

In [2]:
csv_path = "data/raw/>150K/public_150k_plus_211121_add.csv"
csv_path_update = "data/public_150k_plus_220102.csv"

In [3]:
ppp_df = pd.read_csv(csv_path)
ppp_df.head()

Unnamed: 0  LoanNumber DateApproved  SBAOfficeCode ProcessingMethod  \
0           0  9547507704       5/1/20            464              PPP   
1           1  9777677704       5/1/20            464              PPP   
2           2  5791407702       5/1/20           1013              PPP   
3           3  6223567700       5/1/20            920              PPP   
4           4  9662437702       5/1/20            101              PPP   

               BorrowerName        BorrowerAddress      BorrowerCity  \
0     SUMTER COATINGS, INC.  2410 Highway 15 South            Sumter   
1     PLEASANT PLACES, INC.    7684 Southrail Road  North Charleston   
2   BOYER CHILDREN'S CLINIC       1850 BOYER AVE E           SEATTLE   
3  KIRTLEY CONSTRUCTION INC   1661 MARTIN RANCH RD    SAN BERNARDINO   
4              AERO BOX LLC     20101 Cornillie Dr         Roseville   

  BorrowerState  BorrowerZip  ...             BusinessType  \
0            SC      29150.0  ...              Corporation   
1            SC      29420.0  ...      Sole Proprietorship   
2            WA      98112.0  ...  Non-Profit Organization   
3            CA      92407.0  ...              Corporation   
4            MI          NaN  ...                      NaN   

  OriginatingLenderLocationID                      OriginatingLender  \
0                       19248                           Synovus Bank   
1                       19248                           Synovus Bank   
2                        9551  Bank of America, National Association   
3                        9551  Bank of America, National Association   
4                       57328           The Huntington National Bank   

   OriginatingLenderCity  OriginatingLenderState      Gender      Veteran  \
0               COLUMBUS                      GA  Unanswered   Unanswered   
1               COLUMBUS                      GA  Male Owned  Non-Veteran   
2              CHARLOTTE                      NC  Unanswered   Unanswered   
3              CHARLOTTE                      NC  Unanswered   Unanswered   
4               COLUMBUS                      OH  Unanswered   Unanswered   

  NonProfit  ForgivenessAmount ForgivenessDate  
0       NaN          773553.37        11/20/20  
1       NaN          746336.24         8/12/21  
2         Y          696677.49         2/10/21  
3       NaN          395264.11         9/10/21  
4       NaN          370819.35          4/8/21  

[5 rows x 54 columns]

In [4]:
pd.set_option('display.max_rows', 70)
print(ppp_df.isnull().sum())
null_columns=ppp_df.columns[ppp_df.isnull().any()]
df_null = ppp_df[null_columns]
df_null

Unnamed: 0                          0
LoanNumber                          0
DateApproved                        0
SBAOfficeCode                       0
ProcessingMethod                    0
BorrowerName                        4
BorrowerAddress                     6
BorrowerCity                        4
BorrowerState                       0
BorrowerZip                        12
LoanStatusDate                 281932
LoanStatus                          0
Term                                0
SBAGuarantyPercentage               0
InitialApprovalAmount               0
CurrentApprovalAmount               0
UndisbursedAmount                 219
FranchiseName                  933151
ServicingLenderLocationID           0
ServicingLenderName                 0
ServicingLenderAddress              0
ServicingLenderCity                 0
ServicingLenderState                0
ServicingLenderZip                  0
RuralUrbanIndicator                 0
HubzoneIndicator                    0
LMIIndicator

BorrowerName  \
0                                   SUMTER COATINGS, INC.   
1                                   PLEASANT PLACES, INC.   
2                                 BOYER CHILDREN'S CLINIC   
3                                KIRTLEY CONSTRUCTION INC   
4                                            AERO BOX LLC   
...                                                   ...   
968552                           ROY E PAULSON, JR., P.C.   
968553  SWEETWATER COUNTY CHILD DEVELOPMENTAL CENTER, ...   
968554                  ELECTRICAL SYSTEMS OF WYOMING INC   
968555                                     EDEN LIFE CARE   
968556                      S & S JOHNSON ENTERPRISES INC   

                      BorrowerAddress      BorrowerCity  BorrowerZip  \
0               2410 Highway 15 South            Sumter      29150.0   
1                 7684 Southrail Road  North Charleston      29420.0   
2                    1850 BOYER AVE E           SEATTLE      98112.0   
3                1661 MARTIN RANCH RD    SAN BERNARDINO      92407.0   
4                  20101 Cornillie Dr         Roseville          NaN   
...                               ...               ...          ...   
968552                 102 N. Kenwood            CASPER      82601.0   
968553          1715 HITCHING POST DR       GREEN RIVER      82935.0   
968554                   1105 Adon Rd             Rozet      82727.0   
968555  30 N. Gould Street Suite 4000          SHERIDAN      82801.0   
968556           7342 Granite Loop Rd     TETON VILLAGE      83025.0   

       LoanStatusDate  UndisbursedAmount FranchiseName  \
0            12/18/20                0.0           NaN   
1             9/28/21                0.0           NaN   
2             3/17/21                0.0           NaN   
3            10/16/21                0.0           NaN   
4             8/17/21                0.0           NaN   
...               ...                ...           ...   
968552        1/13/21                0.0           NaN   
968553        12/8/20                0.0           NaN   
968554        9/15/21                0.0           NaN   
968555            NaN                0.0           NaN   
968556       10/22/21                0.0           NaN   

                   BusinessAgeDescription       ProjectCity ProjectCountyName  \
0       Existing or more than 2 years old            Sumter            SUMTER   
1       Existing or more than 2 years old  North Charleston        CHARLESTON   
2         New Business or 2 years or less           SEATTLE              KING   
3         New Business or 2 years or less    SAN BERNARDINO    SAN BERNARDINO   
4                              Unanswered               NaN               NaN   
...                                   ...               ...               ...   
968552  Existing or more than 2 years old            CASPER           NATRONA   
968553  Existing or more than 2 years old       GREEN RIVER        SWEETWATER   
968554  Existing or more than 2 years old             Rozet          CAMPBELL   
968555    New Business or 2 years or less          SHERIDAN          SHERIDAN   
968556  Existing or more than 2 years old     TETON VILLAGE             TETON   

        ... PAYROLL_PROCEED MORTGAGE_INTEREST_PROCEED RENT_PROCEED  \
0       ...       769358.78                       NaN          NaN   
1       ...       736927.79                       NaN          NaN   
2       ...       691355.00                       NaN          NaN   
3       ...       499871.00                       NaN          NaN   
4       ...       367437.00                       NaN          NaN   
...     ...             ...                       ...          ...   
968552  ...       150000.00                       NaN          NaN   
968553  ...       145000.00                       NaN          NaN   
968554  ...       149997.00                       NaN          NaN   
968555  ...       150000.00                       NaN          NaN   
968556 

In [5]:
loan_conditions = [
(ppp_df.CurrentApprovalAmount <= 350000) & (ppp_df.CurrentApprovalAmount >= 150000),
(ppp_df.CurrentApprovalAmount <= 2000000) & (ppp_df.CurrentApprovalAmount > 350000),
(ppp_df.CurrentApprovalAmount > 2000000)
]
loan_values = [.05,.03,.01]
ppp_df['Estimated_LenderPercent'] = np.select(loan_conditions, loan_values)
ppp_df

Unnamed: 0  LoanNumber DateApproved  SBAOfficeCode ProcessingMethod  \
0                0  9547507704       5/1/20            464              PPP   
1                1  9777677704       5/1/20            464              PPP   
2                2  5791407702       5/1/20           1013              PPP   
3                3  6223567700       5/1/20            920              PPP   
4                4  9662437702       5/1/20            101              PPP   
...            ...         ...          ...            ...              ...   
968552      968552  4395967002       4/3/20            897              PPP   
968553      968553  6985647108      4/14/20            897              PPP   
968554      968554  7996438405      2/12/21            897              PPS   
968555      968555  9054647103      4/15/20            897              PPP   
968556      968556  9184687004       4/9/20            897              PPP   

                                             BorrowerName  \
0                                   SUMTER COATINGS, INC.   
1                                   PLEASANT PLACES, INC.   
2                                 BOYER CHILDREN'S CLINIC   
3                                KIRTLEY CONSTRUCTION INC   
4                                            AERO BOX LLC   
...                                                   ...   
968552                           ROY E PAULSON, JR., P.C.   
968553  SWEETWATER COUNTY CHILD DEVELOPMENTAL CENTER, ...   
968554                  ELECTRICAL SYSTEMS OF WYOMING INC   
968555                                     EDEN LIFE CARE   
968556                      S & S JOHNSON ENTERPRISES INC   

                      BorrowerAddress      BorrowerCity BorrowerState  \
0               2410 Highway 15 South            Sumter            SC   
1                 7684 Southrail Road  North Charleston            SC   
2                    1850 BOYER AVE E           SEATTLE            WA   
3                1661 MARTIN RANCH RD    SAN BERNARDINO            CA   
4                  20101 Cornillie Dr         Roseville            MI   
...                               ...               ...           ...   
968552                 102 N. Kenwood            CASPER            WY   
968553          1715 HITCHING POST DR       GREEN RIVER            WY   
968554                   1105 Adon Rd             Rozet            WY   
968555  30 N. Gould Street Suite 4000          SHERIDAN            WY   
968556           7342 Granite Loop Rd     TETON VILLAGE            WY   

        BorrowerZip  ... OriginatingLenderLocationID  \
0           29150.0  ...                       19248   
1           29420.0  ...                       19248   
2           98112.0  ...                        9551   
3           92407.0  ...                        9551   
4               NaN  ...                       57328   
...             ...  ...                         ...   
968552      82601.0  ...                       42366   
968553      82935.0  ...                      122813   
968554      82727.0  ...                       77189   
968555      82801.0  ...                       25901   
968556      83025.0  ...                       77193   

                            OriginatingLender  OriginatingLenderCity  \
0                                Synovus Bank               COLUMBUS   
1                                Synovus Bank               COLUMBUS   
2       Bank of America, National Association              CHARLOTTE   
3       Bank of America, National Association              CHARLOTTE   
4                The Huntington National Bank               COLUMBUS   
...                                       ...                    ...   
968552                     Platte Valley Bank             TORRINGTON   
968553                      NebraskaLand Bank           NORTH PLATTE   
968554        First National Bank of Gillette               GILLETTE   
968555                    Small Business Bank           

In [6]:
ppp_df["Estimated_LenderProfit"] = ppp_df['CurrentApprovalAmount'] * ppp_df['Estimated_LenderPercent']
ppp_df['Estimated_LenderPercent'] = pd.Series(["{0:.0f}%".format(val * 100) for val in ppp_df['Estimated_LenderPercent']], index = ppp_df.index)
ppp_df

Unnamed: 0  LoanNumber DateApproved  SBAOfficeCode ProcessingMethod  \
0                0  9547507704       5/1/20            464              PPP   
1                1  9777677704       5/1/20            464              PPP   
2                2  5791407702       5/1/20           1013              PPP   
3                3  6223567700       5/1/20            920              PPP   
4                4  9662437702       5/1/20            101              PPP   
...            ...         ...          ...            ...              ...   
968552      968552  4395967002       4/3/20            897              PPP   
968553      968553  6985647108      4/14/20            897              PPP   
968554      968554  7996438405      2/12/21            897              PPS   
968555      968555  9054647103      4/15/20            897              PPP   
968556      968556  9184687004       4/9/20            897              PPP   

                                             BorrowerName  \
0                                   SUMTER COATINGS, INC.   
1                                   PLEASANT PLACES, INC.   
2                                 BOYER CHILDREN'S CLINIC   
3                                KIRTLEY CONSTRUCTION INC   
4                                            AERO BOX LLC   
...                                                   ...   
968552                           ROY E PAULSON, JR., P.C.   
968553  SWEETWATER COUNTY CHILD DEVELOPMENTAL CENTER, ...   
968554                  ELECTRICAL SYSTEMS OF WYOMING INC   
968555                                     EDEN LIFE CARE   
968556                      S & S JOHNSON ENTERPRISES INC   

                      BorrowerAddress      BorrowerCity BorrowerState  \
0               2410 Highway 15 South            Sumter            SC   
1                 7684 Southrail Road  North Charleston            SC   
2                    1850 BOYER AVE E           SEATTLE            WA   
3                1661 MARTIN RANCH RD    SAN BERNARDINO            CA   
4                  20101 Cornillie Dr         Roseville            MI   
...                               ...               ...           ...   
968552                 102 N. Kenwood            CASPER            WY   
968553          1715 HITCHING POST DR       GREEN RIVER            WY   
968554                   1105 Adon Rd             Rozet            WY   
968555  30 N. Gould Street Suite 4000          SHERIDAN            WY   
968556           7342 Granite Loop Rd     TETON VILLAGE            WY   

        BorrowerZip  ...                      OriginatingLender  \
0           29150.0  ...                           Synovus Bank   
1           29420.0  ...                           Synovus Bank   
2           98112.0  ...  Bank of America, National Association   
3           92407.0  ...  Bank of America, National Association   
4               NaN  ...           The Huntington National Bank   
...             ...  ...                                    ...   
968552      82601.0  ...                     Platte Valley Bank   
968553      82935.0  ...                      NebraskaLand Bank   
968554      82727.0  ...        First National Bank of Gillette   
968555      82801.0  ...                    Small Business Bank   
968556      83025.0  ...                   Bank of Jackson Hole   

       OriginatingLenderCity  OriginatingLenderState        Gender  \
0                   COLUMBUS                      GA    Unanswered   
1                   COLUMBUS                      GA    Male Owned   
2                  CHARLOTTE                      NC    Unanswered   
3                  CHARLOTTE                      NC    Unanswered   
4                   COLUMBUS                      OH    Unanswered   
...                      ...                     ...           ...   
968552            TORRINGTON                      WY    Male Owned   
968553          NORTH PLATTE                      NE    Unanswered   
968554        

In [7]:
ppp_df['DateApproved_str'] = pd.to_datetime(ppp_df['DateApproved'])
ppp_df['DateApproved_ts'] = ppp_df.DateApproved_str.values.astype(np.int64) // 10 ** 9
ppp_df = ppp_df.drop("DateApproved_str", axis=1)
ppp_df

Unnamed: 0  LoanNumber DateApproved  SBAOfficeCode ProcessingMethod  \
0                0  9547507704       5/1/20            464              PPP   
1                1  9777677704       5/1/20            464              PPP   
2                2  5791407702       5/1/20           1013              PPP   
3                3  6223567700       5/1/20            920              PPP   
4                4  9662437702       5/1/20            101              PPP   
...            ...         ...          ...            ...              ...   
968552      968552  4395967002       4/3/20            897              PPP   
968553      968553  6985647108      4/14/20            897              PPP   
968554      968554  7996438405      2/12/21            897              PPS   
968555      968555  9054647103      4/15/20            897              PPP   
968556      968556  9184687004       4/9/20            897              PPP   

                                             BorrowerName  \
0                                   SUMTER COATINGS, INC.   
1                                   PLEASANT PLACES, INC.   
2                                 BOYER CHILDREN'S CLINIC   
3                                KIRTLEY CONSTRUCTION INC   
4                                            AERO BOX LLC   
...                                                   ...   
968552                           ROY E PAULSON, JR., P.C.   
968553  SWEETWATER COUNTY CHILD DEVELOPMENTAL CENTER, ...   
968554                  ELECTRICAL SYSTEMS OF WYOMING INC   
968555                                     EDEN LIFE CARE   
968556                      S & S JOHNSON ENTERPRISES INC   

                      BorrowerAddress      BorrowerCity BorrowerState  \
0               2410 Highway 15 South            Sumter            SC   
1                 7684 Southrail Road  North Charleston            SC   
2                    1850 BOYER AVE E           SEATTLE            WA   
3                1661 MARTIN RANCH RD    SAN BERNARDINO            CA   
4                  20101 Cornillie Dr         Roseville            MI   
...                               ...               ...           ...   
968552                 102 N. Kenwood            CASPER            WY   
968553          1715 HITCHING POST DR       GREEN RIVER            WY   
968554                   1105 Adon Rd             Rozet            WY   
968555  30 N. Gould Street Suite 4000          SHERIDAN            WY   
968556           7342 Granite Loop Rd     TETON VILLAGE            WY   

        BorrowerZip  ... OriginatingLenderCity OriginatingLenderState  \
0           29150.0  ...              COLUMBUS                     GA   
1           29420.0  ...              COLUMBUS                     GA   
2           98112.0  ...             CHARLOTTE                     NC   
3           92407.0  ...             CHARLOTTE                     NC   
4               NaN  ...              COLUMBUS                     OH   
...             ...  ...                   ...                    ...   
968552      82601.0  ...            TORRINGTON                     WY   
968553      82935.0  ...          NORTH PLATTE                     NE   
968554      82727.0  ...              GILLETTE                     WY   
968555      82801.0  ...                LENEXA                     KS   
968556      83025.0  ...               JACKSON                     WY   

              Gender      Veteran  NonProfit  ForgivenessAmount  \
0         Unanswered   Unanswered        NaN          773553.37   
1         Male Owned  Non-Veteran        NaN          746336.24   
2         Unanswered   Unanswered          Y          696677.49   
3         Unanswered   Unanswered        NaN          395264.11   
4         Unanswered   Unanswered        NaN          370819.35   
...              ...          ...        ...                ...   
968552    Male Owned  Non-Veteran        NaN          151037.50   
968553    Unanswered   Unanswered   

In [8]:
ppp_df['ForgivenessDate_str'] = pd.to_datetime(ppp_df['ForgivenessDate'], errors='coerce')
ppp_df['ForgivenessDate_ts'] = ppp_df.ForgivenessDate_str.values.astype(np.int64) // 10 ** 9
ppp_df = ppp_df.drop("ForgivenessDate_str", axis=1)
ppp_df

Unnamed: 0  LoanNumber DateApproved  SBAOfficeCode ProcessingMethod  \
0                0  9547507704       5/1/20            464              PPP   
1                1  9777677704       5/1/20            464              PPP   
2                2  5791407702       5/1/20           1013              PPP   
3                3  6223567700       5/1/20            920              PPP   
4                4  9662437702       5/1/20            101              PPP   
...            ...         ...          ...            ...              ...   
968552      968552  4395967002       4/3/20            897              PPP   
968553      968553  6985647108      4/14/20            897              PPP   
968554      968554  7996438405      2/12/21            897              PPS   
968555      968555  9054647103      4/15/20            897              PPP   
968556      968556  9184687004       4/9/20            897              PPP   

                                             BorrowerName  \
0                                   SUMTER COATINGS, INC.   
1                                   PLEASANT PLACES, INC.   
2                                 BOYER CHILDREN'S CLINIC   
3                                KIRTLEY CONSTRUCTION INC   
4                                            AERO BOX LLC   
...                                                   ...   
968552                           ROY E PAULSON, JR., P.C.   
968553  SWEETWATER COUNTY CHILD DEVELOPMENTAL CENTER, ...   
968554                  ELECTRICAL SYSTEMS OF WYOMING INC   
968555                                     EDEN LIFE CARE   
968556                      S & S JOHNSON ENTERPRISES INC   

                      BorrowerAddress      BorrowerCity BorrowerState  \
0               2410 Highway 15 South            Sumter            SC   
1                 7684 Southrail Road  North Charleston            SC   
2                    1850 BOYER AVE E           SEATTLE            WA   
3                1661 MARTIN RANCH RD    SAN BERNARDINO            CA   
4                  20101 Cornillie Dr         Roseville            MI   
...                               ...               ...           ...   
968552                 102 N. Kenwood            CASPER            WY   
968553          1715 HITCHING POST DR       GREEN RIVER            WY   
968554                   1105 Adon Rd             Rozet            WY   
968555  30 N. Gould Street Suite 4000          SHERIDAN            WY   
968556           7342 Granite Loop Rd     TETON VILLAGE            WY   

        BorrowerZip  ... OriginatingLenderState        Gender      Veteran  \
0           29150.0  ...                     GA    Unanswered   Unanswered   
1           29420.0  ...                     GA    Male Owned  Non-Veteran   
2           98112.0  ...                     NC    Unanswered   Unanswered   
3           92407.0  ...                     NC    Unanswered   Unanswered   
4               NaN  ...                     OH    Unanswered   Unanswered   
...             ...  ...                    ...           ...          ...   
968552      82601.0  ...                     WY    Male Owned  Non-Veteran   
968553      82935.0  ...                     NE    Unanswered   Unanswered   
968554      82727.0  ...                     WY  Female Owned   Unanswered   
968555      82801.0  ...                     KS    Unanswered   Unanswered   
968556      83025.0  ...                     WY    Unanswered   Unanswered   

        NonProfit  ForgivenessAmount  ForgivenessDate  \
0             NaN          773553.37         11/20/20   
1             NaN          746336.24          8/12/21   
2               Y          696677.49          2/10/21   
3             NaN          395264.11          9/10/21   
4             NaN          370819.35           4/8/21   
...           ...                ...              ...   
968552        NaN          151037.50          12/9/20   
968553          Y          150789.04          11/3/20   

In [9]:
pd.set_option('display.max_columns', None)
ppp_df.head()

Unnamed: 0  LoanNumber DateApproved  SBAOfficeCode ProcessingMethod  \
0           0  9547507704       5/1/20            464              PPP   
1           1  9777677704       5/1/20            464              PPP   
2           2  5791407702       5/1/20           1013              PPP   
3           3  6223567700       5/1/20            920              PPP   
4           4  9662437702       5/1/20            101              PPP   

               BorrowerName        BorrowerAddress      BorrowerCity  \
0     SUMTER COATINGS, INC.  2410 Highway 15 South            Sumter   
1     PLEASANT PLACES, INC.    7684 Southrail Road  North Charleston   
2   BOYER CHILDREN'S CLINIC       1850 BOYER AVE E           SEATTLE   
3  KIRTLEY CONSTRUCTION INC   1661 MARTIN RANCH RD    SAN BERNARDINO   
4              AERO BOX LLC     20101 Cornillie Dr         Roseville   

  BorrowerState  BorrowerZip LoanStatusDate    LoanStatus  Term  \
0            SC      29150.0       12/18/20  Paid in Full    24   
1            SC      29420.0        9/28/21  Paid in Full    24   
2            WA      98112.0        3/17/21  Paid in Full    24   
3            CA      92407.0       10/16/21  Paid in Full    24   
4            MI          NaN        8/17/21  Paid in Full    24   

   SBAGuarantyPercentage  InitialApprovalAmount  CurrentApprovalAmount  \
0                    100              769358.78              769358.78   
1                    100              736927.79              736927.79   
2                    100              691355.00              691355.00   
3                    100              499871.00              499871.00   
4                    100              367437.00              367437.00   

   UndisbursedAmount FranchiseName  ServicingLenderLocationID  \
0                0.0           NaN                      19248   
1                0.0           NaN                      19248   
2                0.0           NaN                       9551   
3                0.0           NaN                       9551   
4                0.0           NaN                      57328   

                     ServicingLenderName   ServicingLenderAddress  \
0                           Synovus Bank            1148 Broadway   
1                           Synovus Bank            1148 Broadway   
2  Bank of America, National Association  100 N Tryon St, Ste 170   
3  Bank of America, National Association  100 N Tryon St, Ste 170   
4           The Huntington National Bank            17 S High St.   

  ServicingLenderCity ServicingLenderState ServicingLenderZip  \
0            COLUMBUS                   GA         31901-2429   
1            COLUMBUS                   GA         31901-2429   
2           CHARLOTTE                   NC         28202-4024   
3           CHARLOTTE                   NC         28202-4024   
4            COLUMBUS                   OH         43215-3413   

  RuralUrbanIndicator HubzoneIndicator LMIIndicator  \
0                   U                N            N   
1                   U                Y            Y   
2                   U                N            N   
3                   U                N            N   
4                   U                N            N   

              BusinessAgeDescription       ProjectCity ProjectCountyName  \
0  Existing or more than 2 years old            Sumter            SUMTER   
1  Existing or more than 2 years old  North Charleston        CHARLESTON   
2    New Business or 2 years or less           SEATTLE              KING   
3    New Business or 2 years or less    SAN BERNARDINO    SAN BERNARDINO   
4                         Unanswered               NaN               NaN   

  ProjectState  ProjectZip     CD  JobsReported  NAICSCode        Race  \
0           SC  29150-9662  SC-05          62.0   325510.0  Unanswered   
1           SC  29420-9000  SC-06          73.0   561730.0       White   
2           WA  98112-2922  WA-07          75.0        NaN  Unanswered

conditions = [
    (ppp_df['NA'] == 0),
    (ppp_df['NA'] == 99),
    (ppp_df['NA'] == 72),
    (ppp_df['NA'] == 56),
    (ppp_df['NA'] == 11),
    (ppp_df['NA'] == 71),
    (ppp_df['NA'] == 23),
    (ppp_df['NA'] == 61),
    (ppp_df['NA'] == 52),
    (ppp_df['NA'] == 62),
    (ppp_df['NA'] == 51),
    (ppp_df['NA'] == 55),
    (ppp_df['NA'] == 31),
    (ppp_df['NA'] == 32),
    (ppp_df['NA'] == 33),
    (ppp_df['NA'] == 21),
    (ppp_df['NA'] == 81),
    (ppp_df['NA'] == 54),
    (ppp_df['NA'] == 92),
    (ppp_df['NA'] == 53),
    (ppp_df['NA'] == 44),
    (ppp_df['NA'] == 45),
    (ppp_df['NA'] == 48),
    (ppp_df['NA'] == 49),
    (ppp_df['NA'] == 22),
    (ppp_df['NA'] == 42)
]

values = ['0',
 '99 - Nonclassifiable',
 'Accomodation and Food Services',
 'Administrative and Support and Waste Management and Remediation Services',
 'Agriculture, Forestry, Fishing and Hunting',
 'Arts, Entertainment, and Recreation',
 'Construction',
 'Educational Services',
 'Finance and Insurance',
 'Health Care and Social Assistance',
 'Information',
 'Management of Companies and Enterprises',
 'Manufacturing',
 'Manufacturing',
 'Manufacturing',
 'Mining',
 'Other Services (except Public Administration)',
 'Professional, Scientific, and Technical Services',
 'Public Administration',
 'Real Estate Rental and Leasing',
 'Retail Trade',
 'Retail Trade',
 'Transportation and Warehousing',
 'Transportation and Warehousing',
 'Utilities',
 'Wholesale Trade']

ppp_df['Industry'] = np.select(conditions, values)

ppp_df.head()

In [10]:
naics_ind = pd.read_csv("data/alt_data/NAICSind.csv")
naics_full = pd.read_csv("data/alt_data/NAICSfull.csv")

In [11]:
naics_ind.head()
naics_ind["NA"] = naics_ind["NA"].astype(str)

In [12]:
naics_full["NA"] = naics_full["NAICSCode"].astype(str).str[:2]
naics_full 

NAICSCode                                           LongName  \
0        111110                                    Soybean Farming   
1        111120                  Oilseed (except Soybean) Farming    
2        111130                          Dry Pea and Bean Farming    
3        111140                                      Wheat Farming   
4        111150                                      Corn Farming    
...         ...                                                ...   
1052     926140  Regulation of Agricultural Marketing and Commo...   
1053     926150  Regulation, Licensing, and Inspection of Misce...   
1054     927110                     Space Research and Technology    
1055     928110                                 National Security    
1056     928120                             International Affairs    

     TRIIndustrySector                                  IndustrySubsector  \
0            999 Other                     1111 Oilseed and Grain Farming   
1            999 Other                     1111 Oilseed and Grain Farming   
2            999 Other                     1111 Oilseed and Grain Farming   
3            999 Other                     1111 Oilseed and Grain Farming   
4            999 Other                     1111 Oilseed and Grain Farming   
...                ...                                                ...   
1052         999 Other          9261 Administration of Economic Programs    
1053         999 Other          9261 Administration of Economic Programs    
1054         999 Other                9271 Space Research and Technology    
1055         999 Other  9281 National Security and International Affairs    
1056         999 Other  9281 National Security and International Affairs    

      4DigitNAICS  NA  
0            1111  11  
1            1111  11  
2            1111  11  
3            1111  11  
4            1111  11  
...           ...  ..  
1052         9261  92  
1053         9261  92  
1054         9271  92  
1055         9281  92  
1056         9281  92  

[1057 rows x 6 columns]

In [13]:
naics_full = naics_full.merge(naics_ind, on ="NA", how="left")
naics_full

NAICSCode                                           LongName  \
0        111110                                    Soybean Farming   
1        111120                  Oilseed (except Soybean) Farming    
2        111130                          Dry Pea and Bean Farming    
3        111140                                      Wheat Farming   
4        111150                                      Corn Farming    
...         ...                                                ...   
1052     926140  Regulation of Agricultural Marketing and Commo...   
1053     926150  Regulation, Licensing, and Inspection of Misce...   
1054     927110                     Space Research and Technology    
1055     928110                                 National Security    
1056     928120                             International Affairs    

     TRIIndustrySector                                  IndustrySubsector  \
0            999 Other                     1111 Oilseed and Grain Farming   
1            999 Other                     1111 Oilseed and Grain Farming   
2            999 Other                     1111 Oilseed and Grain Farming   
3            999 Other                     1111 Oilseed and Grain Farming   
4            999 Other                     1111 Oilseed and Grain Farming   
...                ...                                                ...   
1052         999 Other          9261 Administration of Economic Programs    
1053         999 Other          9261 Administration of Economic Programs    
1054         999 Other                9271 Space Research and Technology    
1055         999 Other  9281 National Security and International Affairs    
1056         999 Other  9281 National Security and International Affairs    

      4DigitNAICS  NA                                    Industry  
0            1111  11  Agriculture, Forestry, Fishing and Hunting  
1            1111  11  Agriculture, Forestry, Fishing and Hunting  
2            1111  11  Agriculture, Forestry, Fishing and Hunting  
3            1111  11  Agriculture, Forestry, Fishing and Hunting  
4            1111  11  Agriculture, Forestry, Fishing and Hunting  
...           ...  ..                                         ...  
1052         9261  92                       Public Administration  
1053         9261  92                       Public Administration  
1054         9271  92                       Public Administration  
1055         9281  92                       Public Administration  
1056         9281  92                       Public Administration  

[1057 rows x 7 columns]

In [14]:
ppp_df = ppp_df.merge(naics_full, on="NAICSCode", how="left")
ppp_df.head() #6632

Unnamed: 0  LoanNumber DateApproved  SBAOfficeCode ProcessingMethod  \
0           0  9547507704       5/1/20            464              PPP   
1           1  9777677704       5/1/20            464              PPP   
2           2  5791407702       5/1/20           1013              PPP   
3           3  6223567700       5/1/20            920              PPP   
4           4  9662437702       5/1/20            101              PPP   

               BorrowerName        BorrowerAddress      BorrowerCity  \
0     SUMTER COATINGS, INC.  2410 Highway 15 South            Sumter   
1     PLEASANT PLACES, INC.    7684 Southrail Road  North Charleston   
2   BOYER CHILDREN'S CLINIC       1850 BOYER AVE E           SEATTLE   
3  KIRTLEY CONSTRUCTION INC   1661 MARTIN RANCH RD    SAN BERNARDINO   
4              AERO BOX LLC     20101 Cornillie Dr         Roseville   

  BorrowerState  BorrowerZip LoanStatusDate    LoanStatus  Term  \
0            SC      29150.0       12/18/20  Paid in Full    24   
1            SC      29420.0        9/28/21  Paid in Full    24   
2            WA      98112.0        3/17/21  Paid in Full    24   
3            CA      92407.0       10/16/21  Paid in Full    24   
4            MI          NaN        8/17/21  Paid in Full    24   

   SBAGuarantyPercentage  InitialApprovalAmount  CurrentApprovalAmount  \
0                    100              769358.78              769358.78   
1                    100              736927.79              736927.79   
2                    100              691355.00              691355.00   
3                    100              499871.00              499871.00   
4                    100              367437.00              367437.00   

   UndisbursedAmount FranchiseName  ServicingLenderLocationID  \
0                0.0           NaN                      19248   
1                0.0           NaN                      19248   
2                0.0           NaN                       9551   
3                0.0           NaN                       9551   
4                0.0           NaN                      57328   

                     ServicingLenderName   ServicingLenderAddress  \
0                           Synovus Bank            1148 Broadway   
1                           Synovus Bank            1148 Broadway   
2  Bank of America, National Association  100 N Tryon St, Ste 170   
3  Bank of America, National Association  100 N Tryon St, Ste 170   
4           The Huntington National Bank            17 S High St.   

  ServicingLenderCity ServicingLenderState ServicingLenderZip  \
0            COLUMBUS                   GA         31901-2429   
1            COLUMBUS                   GA         31901-2429   
2           CHARLOTTE                   NC         28202-4024   
3           CHARLOTTE                   NC         28202-4024   
4            COLUMBUS                   OH         43215-3413   

  RuralUrbanIndicator HubzoneIndicator LMIIndicator  \
0                   U                N            N   
1                   U                Y            Y   
2                   U                N            N   
3                   U                N            N   
4                   U                N            N   

              BusinessAgeDescription       ProjectCity ProjectCountyName  \
0  Existing or more than 2 years old            Sumter            SUMTER   
1  Existing or more than 2 years old  North Charleston        CHARLESTON   
2    New Business or 2 years or less           SEATTLE              KING   
3    New Business or 2 years or less    SAN BERNARDINO    SAN BERNARDINO   
4                         Unanswered               NaN               NaN   

  ProjectState  ProjectZip     CD  JobsReported  NAICSCode        Race  \
0           SC  29150-9662  SC-05          62.0   325510.0  Unanswered   
1           SC  29420-9000  SC-06          73.0   561730.0       White   
2           WA  98112-2922  WA-07          75.0        NaN  Unanswered

SBA OFFICE CODE LABELS
0100	  	BOSTON REGIONAL OFFICE
0101	  	MASSACHUSETTS DISTRICT OFFICE
0130	  	SPRINGFIELD BRANCH OFFICE
0150	  	VERMONT DISTRICT OFFICE
0156	  	CONNECTICUT DISTRICT OFFICE
0165	  	RHODE ISLAND DISTRICT OFFICE
0172	  	MAINE DISTRICT OFFICE
0189	  	NEW HAMPSHIRE DISTRICT OFFICE
0200	  	NEW YORK REGIONAL OFFICE
0202	  	NEW YORK DISTRICT OFFICE
0206	  	ELMIRA BRANCH OFFICE
0219	  	ROCHESTER BRANCH OFFICE
0235	  	MELVILLE BRANCH OFFICE
0248	  	SYRACUSE DISTRICT OFFICE
0252	  	PUERTO RICO AND USVI DISTRICT OFFICE
0296	  	BUFFALO DISTRICT OFFICE
0299	  	NEW JERSEY DISTRICT OFFICE
0300	  	PHILADELPHIA REGIONAL OFFICE
0303	  	PHILADELPHIA DISTRICT OFFICE
0304	  	RICHMOND DISTRICT OFFICE
0316	  	HARRISBURG BRANCH OFFICE
0318	  	WILKES BARRE BRANCH OFFICE
0325	  	CHARLESTON BRANCH OFFICE
0341	  	DELAWARE DISTRICT OFFICE
0353	  	WASHINGTON DISTRICT OFFICE
0358	  	PITTSBURGH DISTRICT OFFICE
0373	  	BALTIMORE DISTRICT OFFICE
0390	  	WEST VIRGINIA DISTRICT OFFICE
0400	  	ATLANTA REGIONAL OFFICE
0405	  	GEORGIA DISTRICT OFFICE
0438	  	GULFPORT
0455	  	SOUTH FLORIDA DISTRICT OFFICE
0457	  	KENTUCKY DISTRICT OFFICE
0459	  	ALABAMA DISTRICT OFFICE
0460	  	NORTH CAROLINA DISTRICT OFFICE
0464	  	SOUTH CAROLINA DISTRICT OFFICE
0470	  	MISSISSIPPI DISTRICT OFFICE
0474	  	TENNESSEE DISTRICT OFFICE
0491	  	NORTH FLORIDA DISTRICT OFFICE
0500	  	CHICAGO REGIONAL OFFICE
0507	  	ILLINOIS DISTRICT OFFICE
0508	  	MINNESOTA DISTRICT OFFICE
0515	  	MICHIGAN DISTRICT OFFICE
0517	  	SPRINGFIELD BRANCH OFFICE
0545	  	CINCINNATI BRANCH OFFICE
0547	  	MARQUETTE BRANCH OFFICE
0549	  	CLEVELAND DISTRICT OFFICE
0562	  	INDIANA DISTRICT OFFICE
0563	  	WISCONSIN DISTRICT OFFICE
0593	  	COLUMBUS DISTRICT OFFICE
0600	  	DALLAS REGIONAL OFFICE
0610	  	DALLAS / FT WORTH DISTRICT OFFICE
0637	  	CORPUS CHRISTI BRANCH OFFICE
0639	  	LOWER RIO GRANDE VALLEY DISTRICT OFFICE
0669	  	ARKANSAS DISTRICT OFFICE
0671	  	HOUSTON DISTRICT OFFICE
0677	  	EL PASO DISTRICT OFFICE
0678	  	LUBBOCK DISTRICT OFFICE
0679	  	LOUISIANA DISTRICT OFFICE
0680	  	OKLAHOMA DISTRICT OFFICE
0681	  	SAN ANTONIO DISTRICT OFFICE
0682	  	NEW MEXICO DISTRICT OFFICE
0700	  	KANSAS CITY REGIONAL OFFICE
0709	  	KANSAS CITY DISTRICT OFFICE
0721	  	SPRINGFIELD BRANCH OFFICE
0736	  	CEDAR RAPIDS BRANCH OFFICE
0761	  	DES MOINES DISTRICT OFFICE
0766	  	NEBRASKA DISTRICT OFFICE
0767	  	WICHITA DISTRICT OFFICE
0768	  	ST. LOUIS DISTRICT OFFICE
0800	  	DENVER REGIONAL OFFICE
0811	  	COLORADO DISTRICT OFFICE
0875	  	NORTH DAKOTA DISTRICT OFFICE
0876	  	SOUTH DAKOTA DISTRICT OFFICE
0883	  	UTAH DISTRICT OFFICE
0885	  	MONTANA DISTRICT OFFICE
0897	  	WYOMING DISTRICT OFFICE
0900	  	SAN FRANCISCO REGIONAL OFFICE
0912	  	SAN FRANCISCO DISTRICT OFFICE
0914	  	LOS ANGELES DISTRICT OFFICE
0920	  	SANTA ANA DISTRICT OFFICE
0931	  	SACRAMENTO DISTRICT OFFICE
0942	  	FRESNO DISTRICT OFFICE
0944	  	NEVADA DISTRICT OFFICE
0951	  	HAWAII DISTRICT OFFICE
0954	  	SAN DIEGO DISTRICT OFFICE
0988	  	ARIZONA DISTRICT OFFICE
0995	  	GUAM BRANCH OFFICE
1000	  	SEATTLE REGIONAL OFFICE
1013	  	SEATTLE DISTRICT OFFICE
1084	  	ALASKA DISTRICT OFFICE
1086	  	PORTLAND DISTRICT OFFICE
1087	  	BOISE DISTRICT OFFICE
1094	  	SPOKANE BRANCH OFFICE

In [16]:
conditions = [
(ppp_df['SBAOfficeCode'] ==100),
(ppp_df['SBAOfficeCode'] ==101),
(ppp_df['SBAOfficeCode'] ==130),
(ppp_df['SBAOfficeCode'] ==150),
(ppp_df['SBAOfficeCode'] ==156),
(ppp_df['SBAOfficeCode'] ==165),
(ppp_df['SBAOfficeCode'] ==172),
(ppp_df['SBAOfficeCode'] ==189),
(ppp_df['SBAOfficeCode'] ==200),
(ppp_df['SBAOfficeCode'] ==202),
(ppp_df['SBAOfficeCode'] ==206),
(ppp_df['SBAOfficeCode'] ==219),
(ppp_df['SBAOfficeCode'] ==235),
(ppp_df['SBAOfficeCode'] ==248),
(ppp_df['SBAOfficeCode'] ==252),
(ppp_df['SBAOfficeCode'] ==296),
(ppp_df['SBAOfficeCode'] ==299),
(ppp_df['SBAOfficeCode'] ==300),
(ppp_df['SBAOfficeCode'] ==303),
(ppp_df['SBAOfficeCode'] ==304),
(ppp_df['SBAOfficeCode'] ==316),
(ppp_df['SBAOfficeCode'] ==318),
(ppp_df['SBAOfficeCode'] ==325),
(ppp_df['SBAOfficeCode'] ==341),
(ppp_df['SBAOfficeCode'] ==353),
(ppp_df['SBAOfficeCode'] ==358),
(ppp_df['SBAOfficeCode'] ==373),
(ppp_df['SBAOfficeCode'] ==390),
(ppp_df['SBAOfficeCode'] ==400),
(ppp_df['SBAOfficeCode'] ==405),
(ppp_df['SBAOfficeCode'] ==438),
(ppp_df['SBAOfficeCode'] ==455),
(ppp_df['SBAOfficeCode'] ==457),
(ppp_df['SBAOfficeCode'] ==459),
(ppp_df['SBAOfficeCode'] ==460),
(ppp_df['SBAOfficeCode'] ==464),
(ppp_df['SBAOfficeCode'] ==470),
(ppp_df['SBAOfficeCode'] ==474),
(ppp_df['SBAOfficeCode'] ==491),
(ppp_df['SBAOfficeCode'] ==500),
(ppp_df['SBAOfficeCode'] ==507),
(ppp_df['SBAOfficeCode'] ==508),
(ppp_df['SBAOfficeCode'] ==515),
(ppp_df['SBAOfficeCode'] ==517),
(ppp_df['SBAOfficeCode'] ==545),
(ppp_df['SBAOfficeCode'] ==547),
(ppp_df['SBAOfficeCode'] ==549),
(ppp_df['SBAOfficeCode'] ==562),
(ppp_df['SBAOfficeCode'] ==563),
(ppp_df['SBAOfficeCode'] ==593),
(ppp_df['SBAOfficeCode'] ==600),
(ppp_df['SBAOfficeCode'] ==610),
(ppp_df['SBAOfficeCode'] ==637),
(ppp_df['SBAOfficeCode'] ==639),
(ppp_df['SBAOfficeCode'] ==669),
(ppp_df['SBAOfficeCode'] ==671),
(ppp_df['SBAOfficeCode'] ==677),
(ppp_df['SBAOfficeCode'] ==678),
(ppp_df['SBAOfficeCode'] ==679),
(ppp_df['SBAOfficeCode'] ==680),
(ppp_df['SBAOfficeCode'] ==681),
(ppp_df['SBAOfficeCode'] ==682),
(ppp_df['SBAOfficeCode'] ==700),
(ppp_df['SBAOfficeCode'] ==709),
(ppp_df['SBAOfficeCode'] ==721),
(ppp_df['SBAOfficeCode'] ==736),
(ppp_df['SBAOfficeCode'] ==761),
(ppp_df['SBAOfficeCode'] ==766),
(ppp_df['SBAOfficeCode'] ==767),
(ppp_df['SBAOfficeCode'] ==768),
(ppp_df['SBAOfficeCode'] ==800),
(ppp_df['SBAOfficeCode'] ==811),
(ppp_df['SBAOfficeCode'] ==875),
(ppp_df['SBAOfficeCode'] ==876),
(ppp_df['SBAOfficeCode'] ==883),
(ppp_df['SBAOfficeCode'] ==885),
(ppp_df['SBAOfficeCode'] ==897),
(ppp_df['SBAOfficeCode'] ==900),
(ppp_df['SBAOfficeCode'] ==912),
(ppp_df['SBAOfficeCode'] ==914),
(ppp_df['SBAOfficeCode'] ==920),
(ppp_df['SBAOfficeCode'] ==931),
(ppp_df['SBAOfficeCode'] ==942),
(ppp_df['SBAOfficeCode'] ==944),
(ppp_df['SBAOfficeCode'] ==951),
(ppp_df['SBAOfficeCode'] ==954),
(ppp_df['SBAOfficeCode'] ==988),
(ppp_df['SBAOfficeCode'] ==995),
(ppp_df['SBAOfficeCode'] ==1000),
(ppp_df['SBAOfficeCode'] ==1013),
(ppp_df['SBAOfficeCode'] ==1084),
(ppp_df['SBAOfficeCode'] ==1086),
(ppp_df['SBAOfficeCode'] ==1087),
(ppp_df['SBAOfficeCode'] ==1094),
]

values = [
"BOSTON REGIONAL OFFICE",
"MASSACHUSETTS DISTRICT OFFICE",
"SPRINGFIELD BRANCH OFFICE",
"VERMONT DISTRICT OFFICE",
"CONNECTICUT DISTRICT OFFICE",
"RHODE ISLAND DISTRICT OFFICE",
"MAINE DISTRICT OFFICE",
"NEW HAMPSHIRE DISTRICT OFFICE",
"NEW YORK REGIONAL OFFICE",
"NEW YORK DISTRICT OFFICE",
"ELMIRA BRANCH OFFICE",
"ROCHESTER BRANCH OFFICE",
"MELVILLE BRANCH OFFICE",
"SYRACUSE DISTRICT OFFICE",
"PUERTO RICO AND USVI DISTRICT OFFICE",
"BUFFALO DISTRICT OFFICE",
"NEW JERSEY DISTRICT OFFICE",
"PHILADELPHIA REGIONAL OFFICE",
"PHILADELPHIA DISTRICT OFFICE",
"RICHMOND DISTRICT OFFICE",
"HARRISBURG BRANCH OFFICE",
"WILKES BARRE BRANCH OFFICE",
"CHARLESTON BRANCH OFFICE",
"DELAWARE DISTRICT OFFICE",
"WASHINGTON DISTRICT OFFICE",
"PITTSBURGH DISTRICT OFFICE",
"BALTIMORE DISTRICT OFFICE",
"WEST VIRGINIA DISTRICT OFFICE",
"ATLANTA REGIONAL OFFICE",
"GEORGIA DISTRICT OFFICE",
"GULFPORT",
"SOUTH FLORIDA DISTRICT OFFICE",
"KENTUCKY DISTRICT OFFICE",
"ALABAMA DISTRICT OFFICE",
"NORTH CAROLINA DISTRICT OFFICE",
"SOUTH CAROLINA DISTRICT OFFICE",
"MISSISSIPPI DISTRICT OFFICE",
"TENNESSEE DISTRICT OFFICE",
"NORTH FLORIDA DISTRICT OFFICE",
"CHICAGO REGIONAL OFFICE",
"ILLINOIS DISTRICT OFFICE",
"MINNESOTA DISTRICT OFFICE",
"MICHIGAN DISTRICT OFFICE",
"SPRINGFIELD BRANCH OFFICE",
"CINCINNATI BRANCH OFFICE",
"MARQUETTE BRANCH OFFICE",
"CLEVELAND DISTRICT OFFICE",
"INDIANA DISTRICT OFFICE",
"WISCONSIN DISTRICT OFFICE",
"COLUMBUS DISTRICT OFFICE",
"DALLAS REGIONAL OFFICE",
"DALLAS / FT WORTH DISTRICT OFFICE",
"CORPUS CHRISTI BRANCH OFFICE",
"LOWER RIO GRANDE VALLEY DISTRICT OFFICE",
"ARKANSAS DISTRICT OFFICE",
"HOUSTON DISTRICT OFFICE",
"EL PASO DISTRICT OFFICE",
"LUBBOCK DISTRICT OFFICE",
"LOUISIANA DISTRICT OFFICE",
"OKLAHOMA DISTRICT OFFICE",
"SAN ANTONIO DISTRICT OFFICE",
"NEW MEXICO DISTRICT OFFICE",
"KANSAS CITY REGIONAL OFFICE",
"KANSAS CITY DISTRICT OFFICE",
"SPRINGFIELD BRANCH OFFICE",
"CEDAR RAPIDS BRANCH OFFICE",
"DES MOINES DISTRICT OFFICE",
"NEBRASKA DISTRICT OFFICE",
"WICHITA DISTRICT OFFICE",
"ST. LOUIS DISTRICT OFFICE",
"DENVER REGIONAL OFFICE",
"COLORADO DISTRICT OFFICE",
"NORTH DAKOTA DISTRICT OFFICE",
"SOUTH DAKOTA DISTRICT OFFICE",
"UTAH DISTRICT OFFICE",
"MONTANA DISTRICT OFFICE",
"WYOMING DISTRICT OFFICE",
"SAN FRANCISCO REGIONAL OFFICE",
"SAN FRANCISCO DISTRICT OFFICE",
"LOS ANGELES DISTRICT OFFICE",
"SANTA ANA DISTRICT OFFICE",
"SACRAMENTO DISTRICT OFFICE",
"FRESNO DISTRICT OFFICE",
"NEVADA DISTRICT OFFICE",
"HAWAII DISTRICT OFFICE",
"SAN DIEGO DISTRICT OFFICE",
"ARIZONA DISTRICT OFFICE",
"GUAM BRANCH OFFICE",
"SEATTLE REGIONAL OFFICE",
"SEATTLE DISTRICT OFFICE",
"ALASKA DISTRICT OFFICE",
"PORTLAND DISTRICT OFFICE",
"BOISE DISTRICT OFFICE",
"SPOKANE BRANCH OFFICE"
]
ppp_df['SBAOfficeLabel'] = np.select(conditions, values)

ppp_df.head()

Unnamed: 0  LoanNumber DateApproved  SBAOfficeCode ProcessingMethod  \
0           0  9547507704       5/1/20            464              PPP   
1           1  9777677704       5/1/20            464              PPP   
2           2  5791407702       5/1/20           1013              PPP   
3           3  6223567700       5/1/20            920              PPP   
4           4  9662437702       5/1/20            101              PPP   

               BorrowerName        BorrowerAddress      BorrowerCity  \
0     SUMTER COATINGS, INC.  2410 Highway 15 South            Sumter   
1     PLEASANT PLACES, INC.    7684 Southrail Road  North Charleston   
2   BOYER CHILDREN'S CLINIC       1850 BOYER AVE E           SEATTLE   
3  KIRTLEY CONSTRUCTION INC   1661 MARTIN RANCH RD    SAN BERNARDINO   
4              AERO BOX LLC     20101 Cornillie Dr         Roseville   

  BorrowerState  BorrowerZip LoanStatusDate    LoanStatus  Term  \
0            SC      29150.0       12/18/20  Paid in Full    24   
1            SC      29420.0        9/28/21  Paid in Full    24   
2            WA      98112.0        3/17/21  Paid in Full    24   
3            CA      92407.0       10/16/21  Paid in Full    24   
4            MI          NaN        8/17/21  Paid in Full    24   

   SBAGuarantyPercentage  InitialApprovalAmount  CurrentApprovalAmount  \
0                    100              769358.78              769358.78   
1                    100              736927.79              736927.79   
2                    100              691355.00              691355.00   
3                    100              499871.00              499871.00   
4                    100              367437.00              367437.00   

   UndisbursedAmount FranchiseName  ServicingLenderLocationID  \
0                0.0           NaN                      19248   
1                0.0           NaN                      19248   
2                0.0           NaN                       9551   
3                0.0           NaN                       9551   
4                0.0           NaN                      57328   

                     ServicingLenderName   ServicingLenderAddress  \
0                           Synovus Bank            1148 Broadway   
1                           Synovus Bank            1148 Broadway   
2  Bank of America, National Association  100 N Tryon St, Ste 170   
3  Bank of America, National Association  100 N Tryon St, Ste 170   
4           The Huntington National Bank            17 S High St.   

  ServicingLenderCity ServicingLenderState ServicingLenderZip  \
0            COLUMBUS                   GA         31901-2429   
1            COLUMBUS                   GA         31901-2429   
2           CHARLOTTE                   NC         28202-4024   
3           CHARLOTTE                   NC         28202-4024   
4            COLUMBUS                   OH         43215-3413   

  RuralUrbanIndicator HubzoneIndicator LMIIndicator  \
0                   U                N            N   
1                   U                Y            Y   
2                   U                N            N   
3                   U                N            N   
4                   U                N            N   

              BusinessAgeDescription       ProjectCity ProjectCountyName  \
0  Existing or more than 2 years old            Sumter            SUMTER   
1  Existing or more than 2 years old  North Charleston        CHARLESTON   
2    New Business or 2 years or less           SEATTLE              KING   
3    New Business or 2 years or less    SAN BERNARDINO    SAN BERNARDINO   
4                         Unanswered               NaN               NaN   

  ProjectState  ProjectZip     CD  JobsReported  NAICSCode        Race  \
0           SC  29150-9662  SC-05          62.0   325510.0  Unanswered   
1           SC  29420-9000  SC-06          73.0   561730.0       White   
2           WA  98112-2922  WA-07          75.0        NaN  Unanswered

In [17]:
ppp_df['SBAOfficeLabel'].nunique()

74

In [16]:
loan_conditions = [
(ppp_df.CurrentApprovalAmount <= 350000) & (ppp_df.CurrentApprovalAmount >= 150000),
(ppp_df.CurrentApprovalAmount <= 1000000) & (ppp_df.CurrentApprovalAmount > 350000),
(ppp_df.CurrentApprovalAmount <= 2000000) & (ppp_df.CurrentApprovalAmount > 1000000),
(ppp_df.CurrentApprovalAmount <= 5000000) & (ppp_df.CurrentApprovalAmount > 2000000),
(ppp_df.CurrentApprovalAmount <= 10000000) & (ppp_df.CurrentApprovalAmount > 5000000),  
]
loan_values = ["a $150,000-350,000","b $350,000-1 million", "c $1-2 million","d $2-5 million", "e $5-10 million"]
ppp_df['Loan_Range'] = np.select(loan_conditions, loan_values)
ppp_df.head()

Unnamed: 0  LoanNumber DateApproved  SBAOfficeCode ProcessingMethod  \
0           0  9547507704       5/1/20            464              PPP   
1           1  9777677704       5/1/20            464              PPP   
2           2  5791407702       5/1/20           1013              PPP   
3           3  6223567700       5/1/20            920              PPP   
4           4  9662437702       5/1/20            101              PPP   

               BorrowerName        BorrowerAddress      BorrowerCity  \
0     SUMTER COATINGS, INC.  2410 Highway 15 South            Sumter   
1     PLEASANT PLACES, INC.    7684 Southrail Road  North Charleston   
2   BOYER CHILDREN'S CLINIC       1850 BOYER AVE E           SEATTLE   
3  KIRTLEY CONSTRUCTION INC   1661 MARTIN RANCH RD    SAN BERNARDINO   
4              AERO BOX LLC     20101 Cornillie Dr         Roseville   

  BorrowerState  BorrowerZip LoanStatusDate    LoanStatus  Term  \
0            SC      29150.0       12/18/20  Paid in Full    24   
1            SC      29420.0        9/28/21  Paid in Full    24   
2            WA      98112.0        3/17/21  Paid in Full    24   
3            CA      92407.0       10/16/21  Paid in Full    24   
4            MI          NaN        8/17/21  Paid in Full    24   

   SBAGuarantyPercentage  InitialApprovalAmount  CurrentApprovalAmount  \
0                    100              769358.78              769358.78   
1                    100              736927.79              736927.79   
2                    100              691355.00              691355.00   
3                    100              499871.00              499871.00   
4                    100              367437.00              367437.00   

   UndisbursedAmount FranchiseName  ServicingLenderLocationID  \
0                0.0           NaN                      19248   
1                0.0           NaN                      19248   
2                0.0           NaN                       9551   
3                0.0           NaN                       9551   
4                0.0           NaN                      57328   

                     ServicingLenderName   ServicingLenderAddress  \
0                           Synovus Bank            1148 Broadway   
1                           Synovus Bank            1148 Broadway   
2  Bank of America, National Association  100 N Tryon St, Ste 170   
3  Bank of America, National Association  100 N Tryon St, Ste 170   
4           The Huntington National Bank            17 S High St.   

  ServicingLenderCity ServicingLenderState ServicingLenderZip  \
0            COLUMBUS                   GA         31901-2429   
1            COLUMBUS                   GA         31901-2429   
2           CHARLOTTE                   NC         28202-4024   
3           CHARLOTTE                   NC         28202-4024   
4            COLUMBUS                   OH         43215-3413   

  RuralUrbanIndicator HubzoneIndicator LMIIndicator  \
0                   U                N            N   
1                   U                Y            Y   
2                   U                N            N   
3                   U                N            N   
4                   U                N            N   

              BusinessAgeDescription       ProjectCity ProjectCountyName  \
0  Existing or more than 2 years old            Sumter            SUMTER   
1  Existing or more than 2 years old  North Charleston        CHARLESTON   
2    New Business or 2 years or less           SEATTLE              KING   
3    New Business or 2 years or less    SAN BERNARDINO    SAN BERNARDINO   
4                         Unanswered               NaN               NaN   

  ProjectState  ProjectZip     CD  JobsReported  NAICSCode        Race  \
0           SC  29150-9662  SC-05          62.0   325510.0  Unanswered   
1           SC  29420-9000  SC-06          73.0   561730.0       White   
2           WA  98112-2922  WA-07          75.0        NaN  Unanswered

In [17]:
ppp_df.loc[ppp_df["JobsReported"] == "0", 'JobsReported'] = 1

/opt/anaconda3/lib/python3.8/site-packages/pandas/core/ops/array_ops.py:253: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  res_values = method(rvalues)


In [18]:
#possible use in NAICS division comparison, divisions based on Census of Employment and Wages - BLS - https://data.bls.gov/cew/apps/data_views/data_views.htm#tab=Tables
job_conditions = [
(ppp_df.JobsReported <= 5),
(ppp_df.JobsReported <= 9) & (ppp_df.JobsReported > 5),
(ppp_df.JobsReported <= 19) & (ppp_df.JobsReported > 9),
(ppp_df.JobsReported <= 49) & (ppp_df.JobsReported > 19),
(ppp_df.JobsReported <= 99) & (ppp_df.JobsReported > 49), 
(ppp_df.JobsReported <= 249) & (ppp_df.JobsReported > 99),
(ppp_df.JobsReported <= 499) & (ppp_df.JobsReported > 249),
(ppp_df.JobsReported > 499)
]
job_values = ["Fewer than 5","5 to 9", "10 to 19","20 to 49", "50 to 99", "100 to 249", "250 to 499", "500 or more"]
ppp_df['job_Range'] = np.select(job_conditions, job_values)
ppp_df.head()

Unnamed: 0  LoanNumber DateApproved  SBAOfficeCode ProcessingMethod  \
0           0  9547507704       5/1/20            464              PPP   
1           1  9777677704       5/1/20            464              PPP   
2           2  5791407702       5/1/20           1013              PPP   
3           3  6223567700       5/1/20            920              PPP   
4           4  9662437702       5/1/20            101              PPP   

               BorrowerName        BorrowerAddress      BorrowerCity  \
0     SUMTER COATINGS, INC.  2410 Highway 15 South            Sumter   
1     PLEASANT PLACES, INC.    7684 Southrail Road  North Charleston   
2   BOYER CHILDREN'S CLINIC       1850 BOYER AVE E           SEATTLE   
3  KIRTLEY CONSTRUCTION INC   1661 MARTIN RANCH RD    SAN BERNARDINO   
4              AERO BOX LLC     20101 Cornillie Dr         Roseville   

  BorrowerState  BorrowerZip LoanStatusDate    LoanStatus  Term  \
0            SC      29150.0       12/18/20  Paid in Full    24   
1            SC      29420.0        9/28/21  Paid in Full    24   
2            WA      98112.0        3/17/21  Paid in Full    24   
3            CA      92407.0       10/16/21  Paid in Full    24   
4            MI          NaN        8/17/21  Paid in Full    24   

   SBAGuarantyPercentage  InitialApprovalAmount  CurrentApprovalAmount  \
0                    100              769358.78              769358.78   
1                    100              736927.79              736927.79   
2                    100              691355.00              691355.00   
3                    100              499871.00              499871.00   
4                    100              367437.00              367437.00   

   UndisbursedAmount FranchiseName  ServicingLenderLocationID  \
0                0.0           NaN                      19248   
1                0.0           NaN                      19248   
2                0.0           NaN                       9551   
3                0.0           NaN                       9551   
4                0.0           NaN                      57328   

                     ServicingLenderName   ServicingLenderAddress  \
0                           Synovus Bank            1148 Broadway   
1                           Synovus Bank            1148 Broadway   
2  Bank of America, National Association  100 N Tryon St, Ste 170   
3  Bank of America, National Association  100 N Tryon St, Ste 170   
4           The Huntington National Bank            17 S High St.   

  ServicingLenderCity ServicingLenderState ServicingLenderZip  \
0            COLUMBUS                   GA         31901-2429   
1            COLUMBUS                   GA         31901-2429   
2           CHARLOTTE                   NC         28202-4024   
3           CHARLOTTE                   NC         28202-4024   
4            COLUMBUS                   OH         43215-3413   

  RuralUrbanIndicator HubzoneIndicator LMIIndicator  \
0                   U                N            N   
1                   U                Y            Y   
2                   U                N            N   
3                   U                N            N   
4                   U                N            N   

              BusinessAgeDescription       ProjectCity ProjectCountyName  \
0  Existing or more than 2 years old            Sumter            SUMTER   
1  Existing or more than 2 years old  North Charleston        CHARLESTON   
2    New Business or 2 years or less           SEATTLE              KING   
3    New Business or 2 years or less    SAN BERNARDINO    SAN BERNARDINO   
4                         Unanswered               NaN               NaN   

  ProjectState  ProjectZip     CD  JobsReported  NAICSCode        Race  \
0           SC  29150-9662  SC-05          62.0   325510.0  Unanswered   
1           SC  29420-9000  SC-06          73.0   561730.0       White   
2           WA  98112-2922  WA-07          75.0        NaN  Unanswered

In [19]:
ppp_df.loc[ppp_df["job_Range"] == "0", 'job_Range'] = "Fewer than 5"
ppp_df['job_Range'].value_counts()

20 to 49        417145
10 to 19        242131
50 to 99        161529
100 to 249       84789
5 to 9           31387
250 to 499       22073
Fewer than 5      5145
500 or more       4358
Name: job_Range, dtype: int64

In [20]:
ppp_df['full_add'] = ppp_df.BorrowerAddress + ", " + ppp_df.BorrowerCity + ", " + ppp_df.BorrowerState
ppp_df.head()

Unnamed: 0  LoanNumber DateApproved  SBAOfficeCode ProcessingMethod  \
0           0  9547507704       5/1/20            464              PPP   
1           1  9777677704       5/1/20            464              PPP   
2           2  5791407702       5/1/20           1013              PPP   
3           3  6223567700       5/1/20            920              PPP   
4           4  9662437702       5/1/20            101              PPP   

               BorrowerName        BorrowerAddress      BorrowerCity  \
0     SUMTER COATINGS, INC.  2410 Highway 15 South            Sumter   
1     PLEASANT PLACES, INC.    7684 Southrail Road  North Charleston   
2   BOYER CHILDREN'S CLINIC       1850 BOYER AVE E           SEATTLE   
3  KIRTLEY CONSTRUCTION INC   1661 MARTIN RANCH RD    SAN BERNARDINO   
4              AERO BOX LLC     20101 Cornillie Dr         Roseville   

  BorrowerState  BorrowerZip LoanStatusDate    LoanStatus  Term  \
0            SC      29150.0       12/18/20  Paid in Full    24   
1            SC      29420.0        9/28/21  Paid in Full    24   
2            WA      98112.0        3/17/21  Paid in Full    24   
3            CA      92407.0       10/16/21  Paid in Full    24   
4            MI          NaN        8/17/21  Paid in Full    24   

   SBAGuarantyPercentage  InitialApprovalAmount  CurrentApprovalAmount  \
0                    100              769358.78              769358.78   
1                    100              736927.79              736927.79   
2                    100              691355.00              691355.00   
3                    100              499871.00              499871.00   
4                    100              367437.00              367437.00   

   UndisbursedAmount FranchiseName  ServicingLenderLocationID  \
0                0.0           NaN                      19248   
1                0.0           NaN                      19248   
2                0.0           NaN                       9551   
3                0.0           NaN                       9551   
4                0.0           NaN                      57328   

                     ServicingLenderName   ServicingLenderAddress  \
0                           Synovus Bank            1148 Broadway   
1                           Synovus Bank            1148 Broadway   
2  Bank of America, National Association  100 N Tryon St, Ste 170   
3  Bank of America, National Association  100 N Tryon St, Ste 170   
4           The Huntington National Bank            17 S High St.   

  ServicingLenderCity ServicingLenderState ServicingLenderZip  \
0            COLUMBUS                   GA         31901-2429   
1            COLUMBUS                   GA         31901-2429   
2           CHARLOTTE                   NC         28202-4024   
3           CHARLOTTE                   NC         28202-4024   
4            COLUMBUS                   OH         43215-3413   

  RuralUrbanIndicator HubzoneIndicator LMIIndicator  \
0                   U                N            N   
1                   U                Y            Y   
2                   U                N            N   
3                   U                N            N   
4                   U                N            N   

              BusinessAgeDescription       ProjectCity ProjectCountyName  \
0  Existing or more than 2 years old            Sumter            SUMTER   
1  Existing or more than 2 years old  North Charleston        CHARLESTON   
2    New Business or 2 years or less           SEATTLE              KING   
3    New Business or 2 years or less    SAN BERNARDINO    SAN BERNARDINO   
4                         Unanswered               NaN               NaN   

  ProjectState  ProjectZip     CD  JobsReported  NAICSCode        Race  \
0           SC  29150-9662  SC-05          62.0   325510.0  Unanswered   
1           SC  29420-9000  SC-06          73.0   561730.0       White   
2           WA  98112-2922  WA-07          75.0        NaN  Unanswered

In [21]:
ppp_df["BusinessType"] = ppp_df["BusinessType"].replace({'Limited  Liability Company(LLC)': 'Limited Liability Company(LLC)'})


In [22]:
ppp_df['BusinessAgeDescription'] = ppp_df['BusinessAgeDescription'].fillna('Unanswered')
ppp_df["BusinessType"] = ppp_df["BusinessType"].fillna('Not Available')
ppp_df["NonProfit"] = ppp_df["NonProfit"].fillna('N')
ppp_df.head()

Unnamed: 0  LoanNumber DateApproved  SBAOfficeCode ProcessingMethod  \
0           0  9547507704       5/1/20            464              PPP   
1           1  9777677704       5/1/20            464              PPP   
2           2  5791407702       5/1/20           1013              PPP   
3           3  6223567700       5/1/20            920              PPP   
4           4  9662437702       5/1/20            101              PPP   

               BorrowerName        BorrowerAddress      BorrowerCity  \
0     SUMTER COATINGS, INC.  2410 Highway 15 South            Sumter   
1     PLEASANT PLACES, INC.    7684 Southrail Road  North Charleston   
2   BOYER CHILDREN'S CLINIC       1850 BOYER AVE E           SEATTLE   
3  KIRTLEY CONSTRUCTION INC   1661 MARTIN RANCH RD    SAN BERNARDINO   
4              AERO BOX LLC     20101 Cornillie Dr         Roseville   

  BorrowerState  BorrowerZip LoanStatusDate    LoanStatus  Term  \
0            SC      29150.0       12/18/20  Paid in Full    24   
1            SC      29420.0        9/28/21  Paid in Full    24   
2            WA      98112.0        3/17/21  Paid in Full    24   
3            CA      92407.0       10/16/21  Paid in Full    24   
4            MI          NaN        8/17/21  Paid in Full    24   

   SBAGuarantyPercentage  InitialApprovalAmount  CurrentApprovalAmount  \
0                    100              769358.78              769358.78   
1                    100              736927.79              736927.79   
2                    100              691355.00              691355.00   
3                    100              499871.00              499871.00   
4                    100              367437.00              367437.00   

   UndisbursedAmount FranchiseName  ServicingLenderLocationID  \
0                0.0           NaN                      19248   
1                0.0           NaN                      19248   
2                0.0           NaN                       9551   
3                0.0           NaN                       9551   
4                0.0           NaN                      57328   

                     ServicingLenderName   ServicingLenderAddress  \
0                           Synovus Bank            1148 Broadway   
1                           Synovus Bank            1148 Broadway   
2  Bank of America, National Association  100 N Tryon St, Ste 170   
3  Bank of America, National Association  100 N Tryon St, Ste 170   
4           The Huntington National Bank            17 S High St.   

  ServicingLenderCity ServicingLenderState ServicingLenderZip  \
0            COLUMBUS                   GA         31901-2429   
1            COLUMBUS                   GA         31901-2429   
2           CHARLOTTE                   NC         28202-4024   
3           CHARLOTTE                   NC         28202-4024   
4            COLUMBUS                   OH         43215-3413   

  RuralUrbanIndicator HubzoneIndicator LMIIndicator  \
0                   U                N            N   
1                   U                Y            Y   
2                   U                N            N   
3                   U                N            N   
4                   U                N            N   

              BusinessAgeDescription       ProjectCity ProjectCountyName  \
0  Existing or more than 2 years old            Sumter            SUMTER   
1  Existing or more than 2 years old  North Charleston        CHARLESTON   
2    New Business or 2 years or less           SEATTLE              KING   
3    New Business or 2 years or less    SAN BERNARDINO    SAN BERNARDINO   
4                         Unanswered               NaN               NaN   

  ProjectState  ProjectZip     CD  JobsReported  NAICSCode        Race  \
0           SC  29150-9662  SC-05          62.0   325510.0  Unanswered   
1           SC  29420-9000  SC-06          73.0   561730.0       White   
2           WA  98112-2922  WA-07          75.0        NaN  Unanswered

In [23]:
listbusiness =ppp_df["BusinessType"].value_counts()
listbusiness

Corporation                            415963
Limited Liability Company(LLC)         261542
Subchapter S Corporation               177322
Non-Profit Organization                 54932
Partnership                             18086
Limited Liability Partnership           12737
Sole Proprietorship                     12429
Professional Association                 6780
Cooperative                              2449
501(c)3 – Non Profit                     1790
Non-Profit Childcare Center               909
501(c)6 – Non Profit Membership           765
Not Available                             728
Employee Stock Ownership Plan(ESOP)       636
Self-Employed Individuals                 497
Trust                                     408
Housing Co-op                             152
Joint Venture                             129
Independent Contractors                   116
Single Member LLC                          76
Tribal Concerns                            63
Tenant in Common                  

In [24]:
state_counts = ppp_df["BorrowerState"].value_counts()
print(state_counts)

CA    130622
TX     76185
NY     74113
FL     60041
IL     40815
PA     38537
OH     32650
NJ     32533
MI     29809
MA     27471
GA     25832
WA     23755
VA     23266
NC     22826
CO     19742
MN     19730
MD     18979
WI     17354
IN     16601
MO     16350
TN     15869
AZ     15796
LA     13668
OR     13456
CT     12893
AL     10912
SC     10743
OK      9997
UT      9295
KY      9179
IA      8295
NV      8262
KS      8202
AR      5857
NE      5804
MS      5502
NH      5078
HI      4980
NM      4485
ID      4416
DC      4399
ME      4185
RI      3792
WV      3365
ND      3222
MT      3144
DE      2996
PR      2952
SD      2695
AK      2663
VT      2270
WY      2139
GU       443
VI       290
MP        82
AS        20
Name: BorrowerState, dtype: int64


In [25]:
ppp_df.to_csv("data/ppp_pipe.csv")

In [26]:
ak_df = ppp_df.loc[ppp_df["BorrowerState"] == "AK", :]
al_df = ppp_df.loc[ppp_df["BorrowerState"] == "AL", :]
ar_df = ppp_df.loc[ppp_df["BorrowerState"] == "AR", :]
as_df = ppp_df.loc[ppp_df["BorrowerState"] == "AS", :] #American Samoa
az_df = ppp_df.loc[ppp_df["BorrowerState"] == "AZ", :]
ca_df = ppp_df.loc[ppp_df["BorrowerState"] == "CA", :]
co_df = ppp_df.loc[ppp_df["BorrowerState"] == "CO", :]
ct_df = ppp_df.loc[ppp_df["BorrowerState"] == "CT", :]
dc_df = ppp_df.loc[ppp_df["BorrowerState"] == "DC", :] #D.C.
de_df = ppp_df.loc[ppp_df["BorrowerState"] == "DE", :]
fl_df = ppp_df.loc[ppp_df["BorrowerState"] == "FL", :]
ga_df = ppp_df.loc[ppp_df["BorrowerState"] == "GA", :]
gu_df = ppp_df.loc[ppp_df["BorrowerState"] == "GU", :] #Guam
hi_df = ppp_df.loc[ppp_df["BorrowerState"] == "HI", :]
ia_df = ppp_df.loc[ppp_df["BorrowerState"] == "IA", :]
id_df = ppp_df.loc[ppp_df["BorrowerState"] == "ID", :]
il_df = ppp_df.loc[ppp_df["BorrowerState"] == "IL", :]
in_df = ppp_df.loc[ppp_df["BorrowerState"] == "IN", :]
ks_df = ppp_df.loc[ppp_df["BorrowerState"] == "KS", :]
ky_df = ppp_df.loc[ppp_df["BorrowerState"] == "KY", :]
la_df = ppp_df.loc[ppp_df["BorrowerState"] == "LA", :]
ma_df = ppp_df.loc[ppp_df["BorrowerState"] == "MA", :]
md_df = ppp_df.loc[ppp_df["BorrowerState"] == "MD", :]
me_df = ppp_df.loc[ppp_df["BorrowerState"] == "ME", :]
mi_df = ppp_df.loc[ppp_df["BorrowerState"] == "MI", :]
mn_df = ppp_df.loc[ppp_df["BorrowerState"] == "MN", :]
mo_df = ppp_df.loc[ppp_df["BorrowerState"] == "MO", :]
mp_df = ppp_df.loc[ppp_df["BorrowerState"] == "MP", :] #Saipan Member of Parliament
ms_df = ppp_df.loc[ppp_df["BorrowerState"] == "MS", :]
mt_df = ppp_df.loc[ppp_df["BorrowerState"] == "MT", :]
nc_df = ppp_df.loc[ppp_df["BorrowerState"] == "NC", :]
nd_df = ppp_df.loc[ppp_df["BorrowerState"] == "ND", :]
ne_df = ppp_df.loc[ppp_df["BorrowerState"] == "NE", :]
nh_df = ppp_df.loc[ppp_df["BorrowerState"] == "NH", :]
nj_df = ppp_df.loc[ppp_df["BorrowerState"] == "NJ", :]
nm_df = ppp_df.loc[ppp_df["BorrowerState"] == "NM", :]
nv_df = ppp_df.loc[ppp_df["BorrowerState"] == "NV", :]
ny_df = ppp_df.loc[ppp_df["BorrowerState"] == "NY", :]
oh_df = ppp_df.loc[ppp_df["BorrowerState"] == "OH", :]
ok_df = ppp_df.loc[ppp_df["BorrowerState"] == "OK", :]
or_df = ppp_df.loc[ppp_df["BorrowerState"] == "OR", :]
pa_df = ppp_df.loc[ppp_df["BorrowerState"] == "PA", :]
pr_df = ppp_df.loc[ppp_df["BorrowerState"] == "PR", :] #Puerto Rico
ri_df = ppp_df.loc[ppp_df["BorrowerState"] == "RI", :]
sc_df = ppp_df.loc[ppp_df["BorrowerState"] == "SC", :]
sd_df = ppp_df.loc[ppp_df["BorrowerState"] == "SD", :]
tn_df = ppp_df.loc[ppp_df["BorrowerState"] == "TN", :]
tx_df = ppp_df.loc[ppp_df["BorrowerState"] == "TX", :]
ut_df = ppp_df.loc[ppp_df["BorrowerState"] == "UT", :]
va_df = ppp_df.loc[ppp_df["BorrowerState"] == "VA", :]
vi_df = ppp_df.loc[ppp_df["BorrowerState"] == "VI", :] #Virgin Islands
vt_df = ppp_df.loc[ppp_df["BorrowerState"] == "VT", :]
wa_df = ppp_df.loc[ppp_df["BorrowerState"] == "WA", :]
wi_df = ppp_df.loc[ppp_df["BorrowerState"] == "WI", :]
wv_df = ppp_df.loc[ppp_df["BorrowerState"] == "WV", :]
wy_df = ppp_df.loc[ppp_df["BorrowerState"] == "WY", :]

In [27]:
ak_df.to_csv("data/state_data/pipe_states/ak.csv") 
al_df.to_csv("data/state_data/pipe_states/al.csv")
ar_df.to_csv("data/state_data/pipe_states/ar.csv") 
as_df.to_csv("data/state_data/pipe_states/as.csv")
az_df.to_csv("data/state_data/pipe_states/az.csv")  
ca_df.to_csv("data/state_data/pipe_states/ca.csv")  
co_df.to_csv("data/state_data/pipe_states/co.csv") 
ct_df.to_csv("data/state_data/pipe_states/ct.csv")  
dc_df.to_csv("data/state_data/pipe_states/dc.csv")
de_df.to_csv("data/state_data/pipe_states/de.csv") 
fl_df.to_csv("data/state_data/pipe_states/fl.csv") 
ga_df.to_csv("data/state_data/pipe_states/ga.csv") 
gu_df.to_csv("data/state_data/pipe_states/gu.csv")
hi_df.to_csv("data/state_data/pipe_states/hi.csv") 
ia_df.to_csv("data/state_data/pipe_states/ia.csv")  
id_df.to_csv("data/state_data/pipe_states/id.csv")  
il_df.to_csv("data/state_data/pipe_states/il.csv")  
in_df.to_csv("data/state_data/pipe_states/in.csv") 
ks_df.to_csv("data/state_data/pipe_states/ks.csv")  
ky_df.to_csv("data/state_data/pipe_states/ky.csv")  
la_df.to_csv("data/state_data/pipe_states/la.csv")
ma_df.to_csv("data/state_data/pipe_states/ma.csv") 
md_df.to_csv("data/state_data/pipe_states/md.csv")
me_df.to_csv("data/state_data/pipe_states/me.csv")  
mi_df.to_csv("data/state_data/pipe_states/mi.csv") 
mn_df.to_csv("data/state_data/pipe_states/mn.csv") 
mo_df.to_csv("data/state_data/pipe_states/mo.csv")  
mp_df.to_csv("data/state_data/pipe_states/mp.csv") 
ms_df.to_csv("data/state_data/pipe_states/ms.csv")
mt_df.to_csv("data/state_data/pipe_states/mt.csv")
nc_df.to_csv("data/state_data/pipe_states/nc.csv") 
nd_df.to_csv("data/state_data/pipe_states/nd.csv")  
ne_df.to_csv("data/state_data/pipe_states/ne.csv")
nh_df.to_csv("data/state_data/pipe_states/nh.csv")  
nj_df.to_csv("data/state_data/pipe_states/nj.csv")
nm_df.to_csv("data/state_data/pipe_states/nm.csv") 
nv_df.to_csv("data/state_data/pipe_states/nv.csv")  
ny_df.to_csv("data/state_data/pipe_states/ny.csv")  
oh_df.to_csv("data/state_data/pipe_states/oh.csv")
ok_df.to_csv("data/state_data/pipe_states/ok.csv") 
or_df.to_csv("data/state_data/pipe_states/or.csv")  
pa_df.to_csv("data/state_data/pipe_states/pa.csv")
pr_df.to_csv("data/state_data/pipe_states/pr.csv")  
ri_df.to_csv("data/state_data/pipe_states/ri.csv") 
sc_df.to_csv("data/state_data/pipe_states/sc.csv") 
sd_df.to_csv("data/state_data/pipe_states/sd.csv") 
tn_df.to_csv("data/state_data/pipe_states/tn.csv")  
tx_df.to_csv("data/state_data/pipe_states/tx.csv")  
ut_df.to_csv("data/state_data/pipe_states/ut.csv")  
va_df.to_csv("data/state_data/pipe_states/va.csv")  
vi_df.to_csv("data/state_data/pipe_states/vi.csv")  
vt_df.to_csv("data/state_data/pipe_states/vt.csv")
wa_df.to_csv("data/state_data/pipe_states/wa.csv")  
wi_df.to_csv("data/state_data/pipe_states/wi.csv")  
wv_df.to_csv("data/state_data/pipe_states/wv.csv") 
wy_df.to_csv("data/state_data/pipe_states/wy.csv")  

Possible Additions to Pipeline:<br>
Location-based:    http://www.city-data.com/ <br>
Dates:
using deep learning with time series data is to split the date into multiple categorical variables (year, month, week, day of week, day of month, and Booleans for whether it’s the start/end of a month/quarter/year). 